examples/adc.rs

```diff
@@ -0,0 +1,90 @@
+//! Example of reading a voltage with ADC1
+//!
+//! For an example of using ADC1 and ADC2 together, see examples/adc12.rs
+
+#![no_main]
+#![no_std]
+
+use cortex_m_rt::entry;
+
+use embedded_hal_02::adc::OneShot;
+use stm32h5xx_hal::{
+    adc, delay::Delay, pac, prelude::*, rcc::rec::AdcDacClkSel,
+};
+use utilities::logger::info;
+
+#[macro_use]
+mod utilities;
+
+#[entry]
+fn main() -> ! {
+    utilities::logger::init();
+    let cp = cortex_m::Peripherals::take().unwrap();
+    let dp = pac::Peripherals::take().unwrap();
+
+    // Constrain and Freeze power
+    info!("Setup PWR...                  ");
+    let pwr = dp.PWR.constrain();
+    let pwrcfg = pwr.freeze();
+
+    // Constrain and Freeze clock
+    info!("Setup RCC...                  ");
+    let rcc = dp.RCC.constrain();
+
+    // We need to configure a clock for adc_ker_ck_input. The default
+    // adc_ker_ck_input is pll2_p_ck, but we will use per_ck. per_ck is sourced
+    // from the 64MHz HSI
+    //
+    // adc_ker_ck_input is then divided by the ADC prescaler to give f_adc. The
+    // maximum f_adc is 50MHz
+    let mut ccdr = rcc
+        .sys_ck(192.MHz())
+        .pll1_q_ck(64.MHz())
+        .freeze(&pwrcfg, &dp.SBS);
+
+    // Switch adc_ker_ck_input multiplexer to per_ck
+    ccdr.peripheral.kernel_adcdac_clk_mux(AdcDacClkSel::HsiKer);
+
+    info!("");
+    info!("stm32h5xx-hal example - ADC");
+    info!("");
+
+    let mut delay = Delay::new(cp.SYST, &ccdr.clocks);
+
+    // Setup ADC
+    let mut adc1 = adc::Adc::new(
+        dp.ADC1,
+        4.MHz(),
+        &mut delay,
+        ccdr.peripheral.ADC,
+        &ccdr.clocks,
+        &pwrcfg,
+    );
```

Comment:
I am not too happy with the api for setting up the adc. I think I would prefer something more like the [G4 hal](https://github.com/stm32-rs/stm32g4xx-hal/blob/main/examples/adc-one-shot.rs#L37)

```rust
let adc12_common = dp.ADC12_COMMON.claim(Default::default(), &mut rcc);
let mut adc = adc12_common.claim_and_configure(
    dp.ADC2,
    stm32g4xx_hal::adc::config::AdcConfig::default(),
    &mut delay,
);
```

Here settings are divided into the individual config and the common config. Config structs are used with sensible defaults for those that do not care instead of individual arguments for frequency etc/or setter methods that have to be used after the fact

Reply:
Probably, I just have not really changed anything from H7 yet. But yes that would probably be more consistent with the rest of the H5 hal.

The H503 only has a ADC1 so it does not have the ADCC common register block as a separate block so that would probably not work with the above from G4